File: src/Entity/FightStatistic.php
<?php
namespace App\Entity;

use Doctrine\ORM\Mapping as ORM;
use App\Repository\FightStatisticRepository;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity(repositoryClass: FightStatisticRepository::class)]
#[ORM\Table(name: 'fight_statistic')]
class FightStatistic
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\ManyToOne(targetEntity: FightResult::class)]
    #[ORM\JoinColumn(name: "fight_result_id", referencedColumnName: "resultId", nullable: false)]
    private ?FightResult $fightResult = null;

    #[ORM\ManyToOne(targetEntity: Fighter::class)]
    #[ORM\JoinColumn(name: "fighter_id", referencedColumnName: "fighterId", nullable: false)]
    private ?Fighter $fighter = null;

    #[ORM\Column(name: 'round', type: 'integer', nullable: true)]
    #[Assert\Range(min: 1, max: 12)]
    private ?int $round = null;

    #[ORM\Column(name: 'punches_landed', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $punchesLanded = 0;

    #[ORM\Column(name: 'punches_thrown', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $punchesThrown = 0;

    #[ORM\Column(name: 'body_shots_landed', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $bodyShotsLanded = 0;

    #[ORM\Column(name: 'body_jabs_landed', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $bodyJabsLanded = 0;

    #[ORM\Column(name: 'body_power_landed', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $bodyPowerLanded = 0;

    #[ORM\Column(name: 'jabs_landed', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $jabsLanded = 0;

    #[ORM\Column(name: 'jabs_thrown', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $jabsThrown = 0;

    #[ORM\Column(name: 'power_punches_landed', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $powerPunchesLanded = 0;

    #[ORM\Column(name: 'power_punches_thrown', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $powerPunchesThrown = 0;

    #[ORM\Column(name: 'uppercuts_landed', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $uppercutsLanded = 0;

    #[ORM\Column(name: 'uppercuts_thrown', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $uppercutsThrown = 0;

    #[ORM\Column(name: 'right_hand_landed', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $rightHandLanded = 0;

    #[ORM\Column(name: 'right_hand_thrown', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $rightHandThrown = 0;

    #[ORM\Column(name: 'left_hand_landed', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $leftHandLanded = 0;

    #[ORM\Column(name: 'left_hand_thrown', type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $leftHandThrown = 0;

    #[ORM\Column(type: 'integer', options: ['default' => 0])]
    #[Assert\PositiveOrZero]
    private int $knockdowns = 0;

    #[ORM\Column(type: 'text', nullable: true)]
    private ?string $commentary = null;

    public function getId(): ?int { return $this->id; }

    public function getFightResult(): ?FightResult { return $this->fightResult; }
    public function setFightResult(?FightResult $v): self { $this->fightResult = $v; return $this; }

    public function getFighter(): ?Fighter { return $this->fighter; }
    public function setFighter(?Fighter $v): self { $this->fighter = $v; return $this; }

    public function getRound(): ?int { return $this->round; }
    public function setRound(?int $v): self { $this->round = $v; return $this; }

    public function getPunchesLanded(): int { return $this->punchesLanded; }
    public function setPunchesLanded(int $v): self { $this->punchesLanded = $v; return $this; }

    public function getPunchesThrown(): int { return $this->punchesThrown; }
    public function setPunchesThrown(int $v): self { $this->punchesThrown = $v; return $this; }

    public function getBodyShotsLanded(): int { return $this->bodyShotsLanded; }
    public function setBodyShotsLanded(int $v): self { $this->bodyShotsLanded = $v; return $this; }

    public function getBodyJabsLanded(): int { return $this->bodyJabsLanded; }
    public function setBodyJabsLanded(int $v): self { $this->bodyJabsLanded = $v; return $this; }

    public function getBodyPowerLanded(): int { return $this->bodyPowerLanded; }
    public function setBodyPowerLanded(int $v): self { $this->bodyPowerLanded = $v; return $this; }

    public function getJabsLanded(): int { return $this->jabsLanded; }
    public function setJabsLanded(int $v): self { $this->jabsLanded = $v; return $this; }

    public function getJabsThrown(): int { return $this->jabsThrown; }
    public function setJabsThrown(int $v): self { $this->jabsThrown = $v; return $this; }

    public function getPowerPunchesLanded(): int { return $this->powerPunchesLanded; }
    public function setPowerPunchesLanded(int $v): self { $this->powerPunchesLanded = $v; return $this; }

    public function getPowerPunchesThrown(): int { return $this->powerPunchesThrown; }
    public function setPowerPunchesThrown(int $v): self { $this->powerPunchesThrown = $v; return $this; }

    public function getUppercutsLanded(): int { return $this->uppercutsLanded; }
    public function setUppercutsLanded(int $v): self { $this->uppercutsLanded = $v; return $this; }

    public function getUppercutsThrown(): int { return $this->uppercutsThrown; }
    public function setUppercutsThrown(int $v): self { $this->uppercutsThrown = $v; return $this; }

    public function getRightHandLanded(): int { return $this->rightHandLanded; }
    public function setRightHandLanded(int $v): self { $this->rightHandLanded = $v; return $this; }

    public function getRightHandThrown(): int { return $this->rightHandThrown; }
    public function setRightHandThrown(int $v): self { $this->rightHandThrown = $v; return $this; }

    public function getLeftHandLanded(): int { return $this->leftHandLanded; }
    public function setLeftHandLanded(int $v): self { $this->leftHandLanded = $v; return $this; }

    public function getLeftHandThrown(): int { return $this->leftHandThrown; }
    public function setLeftHandThrown(int $v): self { $this->leftHandThrown = $v; return $this; }

    public function getKnockdowns(): int { return $this->knockdowns; }
    public function setKnockdowns(int $v): self { $this->knockdowns = $v; return $this; }

    public function getCommentary(): ?string { return $this->commentary; }
    public function setCommentary(?string $v): self { $this->commentary = $v; return $this; }

    public function getPunchAccuracy(): float
    {
        if ($this->punchesThrown == 0) return 0.0;
        return ($this->punchesLanded * 100.0) / $this->punchesThrown;
    }

    public function getJabAccuracy(): float
    {
        if ($this->jabsThrown == 0) return 0.0;
        return ($this->jabsLanded * 100.0) / $this->jabsThrown;
    }

    public function getPowerAccuracy(): float
    {
        if ($this->powerPunchesThrown == 0) return 0.0;
        return ($this->powerPunchesLanded * 100.0) / $this->powerPunchesThrown;
    }

    public function getUppercutAccuracy(): float
    {
        if ($this->uppercutsThrown == 0) return 0.0;
        return ($this->uppercutsLanded * 100.0) / $this->uppercutsThrown;
    }

    public function getRightHandAccuracy(): float
    {
        if ($this->rightHandThrown == 0) return 0.0;
        return ($this->rightHandLanded * 100.0) / $this->rightHandThrown;
    }

    public function getLeftHandAccuracy(): float
    {
        if ($this->leftHandThrown == 0) return 0.0;
        return ($this->leftHandLanded * 100.0) / $this->leftHandThrown;
    }

    #[Assert\IsTrue(message: 'Landed punches cannot exceed thrown punches')]
    public function isLandedValid(): bool
    {
        return $this->punchesLanded <= $this->punchesThrown
            && $this->jabsLanded <= $this->jabsThrown
            && $this->powerPunchesLanded <= $this->powerPunchesThrown
            && $this->uppercutsLanded <= $this->uppercutsThrown
            && $this->rightHandLanded <= $this->rightHandThrown
            && $this->leftHandLanded <= $this->leftHandThrown;
    }

    #[Assert\IsTrue(message: 'Total punches thrown must equal right hand thrown + left hand thrown')]
    public function isHandThrownLogicValid(): bool
    {
        return $this->punchesThrown === ($this->rightHandThrown + $this->leftHandThrown);
    }

    #[Assert\IsTrue(message: 'Total punches landed must equal right hand landed + left hand landed')]
    public function isHandLandedLogicValid(): bool
    {
        return $this->punchesLanded === ($this->rightHandLanded + $this->leftHandLanded);
    }

    #[Assert\IsTrue(message: 'Uppercuts thrown cannot exceed total punches thrown')]
    public function isUppercutLogicValid(): bool
    {
        return $this->uppercutsThrown <= $this->punchesThrown && $this->uppercutsLanded <= $this->punchesLanded;
    }
}

File: src/main/java/tn/smartfight/model/FightStatistic.java
package tn.smartfight.model;

public class FightStatistic {
    private int id;
    private int fightId;
    private int fighterId;
    private Integer roundNumber; // null = aggregate row
    private int punchesThrown, punchesLanded;
    private int rightHandThrown, rightHandLanded;
    private int leftHandThrown, leftHandLanded;
    private int powerPunchesThrown, powerPunchesLanded;
    private int jabsThrown, jabsLanded;
    private int uppercutsThrown, uppercutsLanded;
    private int bodyShotsLanded;
    private int knockdowns;
    private String commentary;
    private String fighterName;

    public FightStatistic() {}

    public int getId() { return id; }
    public void setId(int id) { this.id = id; }

    public int getFightId() { return fightId; }
    public void setFightId(int fightId) { this.fightId = fightId; }

    public int getFighterId() { return fighterId; }
    public void setFighterId(int fighterId) { this.fighterId = fighterId; }

    public Integer getRoundNumber() { return roundNumber; }
    public void setRoundNumber(Integer roundNumber) { this.roundNumber = roundNumber; }

    public int getPunchesThrown() { return punchesThrown; }
    public void setPunchesThrown(int v) { this.punchesThrown = v; }

    public int getPunchesLanded() { return punchesLanded; }
    public void setPunchesLanded(int v) { this.punchesLanded = v; }

    public int getRightHandThrown() { return rightHandThrown; }
    public void setRightHandThrown(int v) { this.rightHandThrown = v; }

    public int getRightHandLanded() { return rightHandLanded; }
    public void setRightHandLanded(int v) { this.rightHandLanded = v; }

    public int getLeftHandThrown() { return leftHandThrown; }
    public void setLeftHandThrown(int v) { this.leftHandThrown = v; }

    public int getLeftHandLanded() { return leftHandLanded; }
    public void setLeftHandLanded(int v) { this.leftHandLanded = v; }

    public int getPowerPunchesThrown() { return powerPunchesThrown; }
    public void setPowerPunchesThrown(int v) { this.powerPunchesThrown = v; }

    public int getPowerPunchesLanded() { return powerPunchesLanded; }
    public void setPowerPunchesLanded(int v) { this.powerPunchesLanded = v; }

    public int getJabsThrown() { return jabsThrown; }
    public void setJabsThrown(int v) { this.jabsThrown = v; }

    public int getJabsLanded() { return jabsLanded; }
    public void setJabsLanded(int v) { this.jabsLanded = v; }

    public int getUppercutsThrown() { return uppercutsThrown; }
    public void setUppercutsThrown(int v) { this.uppercutsThrown = v; }

    public int getUppercutsLanded() { return uppercutsLanded; }
    public void setUppercutsLanded(int v) { this.uppercutsLanded = v; }

    public int getBodyShotsLanded() { return bodyShotsLanded; }
    public void setBodyShotsLanded(int v) { this.bodyShotsLanded = v; }

    public int getKnockdowns() { return knockdowns; }
    public void setKnockdowns(int v) { this.knockdowns = v; }

    public String getCommentary() { return commentary; }
    public void setCommentary(String commentary) { this.commentary = commentary; }

    public String getFighterName() { return fighterName; }
    public void setFighterName(String fighterName) { this.fighterName = fighterName; }

    public double getPunchAccuracy() {
        return punchesThrown > 0 ? punchesLanded * 100.0 / punchesThrown : 0;
    }

    public double getPowerAccuracy() {
        return powerPunchesThrown > 0 ? powerPunchesLanded * 100.0 / powerPunchesThrown : 0;
    }

    public double getJabAccuracy() {
        return jabsThrown > 0 ? jabsLanded * 100.0 / jabsThrown : 0;
    }

    public double getUppercutAccuracy() {
        return uppercutsThrown > 0 ? uppercutsLanded * 100.0 / uppercutsThrown : 0;
    }
}
